lib/view_model/active_record/cache.rb

```diff
@@ -35,85 +53,98 @@ def clear
 
-    # Load existing serializations from the cache
-    cached_serializations = worker.load_from_cache(self, ids)
-    cached_serializations.each do |id, data|
-      positions[id].each do |idx|
-        data_serializations[idx] = data
-      end
+    def initialize(migration_versions:, serialize_context:)
+      @worklist                = {} # Hash[type_name, Hash[id, WorklistEntry]]
+      @resolved_references     = {} # Hash[refname, json]
+      @migration_versions      = migration_versions
+      @migrated_cache_versions = {}
+      @serialize_context       = serialize_context
     end
 
-    # Resolve and serialize missing views
-    missing_ids = ids.to_set.subtract(cached_serializations.keys)
+    def render_from_cache(viewmodel_class, ids, initial_viewmodels: nil, locked: false)
+      viewmodel_class.transaction do
+        root_serializations = Array.new(ids.size)
 
-    # If initial viewmodels have been locked, we can serialize them for cache
-    #  misses.
-    available_viewmodels =
-      if locked
-        initial_viewmodels&.each_with_object({}) do |vm, h|
-          h[vm.id] = vm if missing_ids.include?(vm.id)
+        # Collect input array positions for each id, allowing duplicates
+        positions = ids.each_with_index.with_object({}) do |(id, i), h|
+          (h[id] ||= []) << i
         end
-      end
 
-    @viewmodel_class.transaction do
-      # Load remaining views and serialize
-      viewmodels = worker.find_and_preload_viewmodels(@viewmodel_class, missing_ids.to_a,
-                                                      available_viewmodels: available_viewmodels)
+        # If duplicates are specified, fetch each only once
+        ids = positions.keys
+
+        ids_to_render = ids.to_set
 
-      loaded_serializations = worker.serialize_and_cache(viewmodels)
-      loaded_serializations.each do |id, data|
-        positions[id].each do |idx|
-          data_serializations[idx] = data
+        if viewmodel_class < CacheableView
+          # Load existing serializations from the cache
+          cached_serializations = load_from_cache(viewmodel_class.viewmodel_cache, ids)
+          cached_serializations.each do |id, data|
+            positions[id].each do |idx|
+              root_serializations[idx] = data
+            end
+          end
+
+          ids_to_render.subtract(cached_serializations.keys)
+
+          # If initial root viewmodels were provided, call hooks on any
+          # viewmodels which were rendered from the cache to ensure that the
+          # root is visible (in isolation). Other than this, no traversal
+          # callbacks are performed for cache-rendered views. This particularly
+          # requires care for references: if a visible view may refer to
+          # non-visible cacheable views, those referenced views will not be
+          # access control checked.
+          initial_viewmodels&.each do |v|
+            next unless cached_serializations.has_key?(v.id)
```

Comment:
It's been a while since I looked at this, but this looks like quite a departure from the current implementation. Justify why the presence of a cached serialization is sufficient to skip the access control check? In the previous implementation they're always checked, which effectively means "you can use their cached views if you can see the root", so access controls _always_ apply. In the new implementation the meaning is "you can use cached views if they have been cached", which sometimes skips access controls.

However I believe it's correct to only check cacheable views, since non-cached views will always be checked as part of rendering.

Reply:
Yeah. Previously the callbacks were double-invoked in the case of a cache miss, which seemed undesirable.